File: packages/tutanota-crypto/test/HmacTest.ts
import o from "@tutao/otest"
import { aes256RandomKey, hmacSha256, verifyHmacSha256 } from "../lib/index.js"
import { assertThrows } from "@tutao/tutanota-test-utils"
import { CryptoError } from "../lib/misc/CryptoError.js"

o.spec("hmac", function () {
	o("round trip", function () {
		const key = aes256RandomKey()
		const data = new Uint8Array([0, 1, 2, 3, 4, 5, 6])
		const tag = hmacSha256(key, data)
		verifyHmacSha256(key, data, tag)
	})
	o("throws if data is not the same", async function () {
		const key = aes256RandomKey()
		const data = new Uint8Array([0, 1, 2, 3, 4, 5, 6])
		const badData = new Uint8Array([6, 5, 4, 3, 2, 1, 0])
		const tag = hmacSha256(key, data)
		await assertThrows(CryptoError, async () => verifyHmacSha256(key, badData, tag))
	})
	o("throws if key is not the same", async function () {
		const key = aes256RandomKey()
		const badKey = aes256RandomKey()
		const data = new Uint8Array([0, 1, 2, 3, 4, 5, 6])
		const tag = hmacSha256(key, data)
		await assertThrows(CryptoError, async () => verifyHmacSha256(badKey, data, tag))
	})
})

File: app-android/app/src/androidTest/java/de/tutao/tutanota/HmacTest.kt
package de.tutao.tutanota

import androidx.test.ext.junit.runners.AndroidJUnit4
import de.tutao.tutashared.CryptoError
import de.tutao.tutashared.crypto.Crypto
import org.junit.Assert.assertThrows
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.security.SecureRandom

@RunWith(AndroidJUnit4::class)
class HmacTest {
	lateinit var randomizer: SecureRandom
	lateinit var key: ByteArray
	lateinit var data: ByteArray
	lateinit var macTag: ByteArray

	@Before
	fun setup() {
		randomizer = SecureRandom()
		key = ByteArray(32)
		data = ByteArray(256)

		randomizer.nextBytes(key)
		randomizer.nextBytes(data)

		macTag = Crypto.hmacSha256(key, data)
	}

	@Test
	fun roundTrip() {
		Crypto.verifyHmacSha256(key, data, macTag)
	}

	@Test
	fun badKey() {
		val badKey = ByteArray(32)
		randomizer.nextBytes(badKey)

		assertThrows(CryptoError::class.java) {
			Crypto.verifyHmacSha256(badKey, data, macTag)
		}
	}

	@Test
	fun badData() {
		val badData = ByteArray(256)
		randomizer.nextBytes(badData)

		assertThrows(CryptoError::class.java) {
			Crypto.verifyHmacSha256(key, badData, macTag)
		}
	}
}
